### DISCUSSION TITLE
"AHT2x" and "AHT sensor" labeled boards problem

### DISCUSSION THREAD

**Discussion by xostia** [arendst/Tasmota#21088] [Q&A]
Hi,

We are working in a device that combines the TSL2591 light sensor and AHT2x for temperature and humidity. AHT2x are compatible with other I2C devices in the same bus (AHT1x not) ... but I do not understand what version is "AHT Sensor".

Last units of AHT sensor that I received, do not have "AHT2x" written in the back but "AHT Sensor", even when in the picture of the seller it appears with the "AHT2x", :-(

The problem is that the ones with "AHT2x" work perfect but the others with "AHT Sensor" are not detected. There are versions with AHT21 (black) or AHT20 (silver) sensor chip, but any one works in the "AHT2x" boards but none of them in "AHT Sensor" boards.

Anybody have an explanation? ... it seems impossible to buy the AHT2x now, I bought from 3 different sellers and they send always the "AHT Sensor" labeled.

I include a picture (I only have here one AHT2x with AHT21 sensor but I used the AHT20 silver and also works).

Thank you.

![AHT2x](https://github.com/arendst/Tasmota/assets/90287008/9bc134e5-1a46-4b47-9a73-7bb096451155)
![AHT Sensor](https://github.com/arendst/Tasmota/assets/90287008/b951ee1e-b06d-4570-b220-9e632981090d)

**Comment by sfromis** [comment]
There exists no "AHT2x" sensor. You may have a board designed to accommodate more than one of AHT20/AHT21/AHT25, but that's different sensors, and they did not bother with separate silk screens per chip variant. You could say that printing just "AHT Sensor" is "less misleading", in both cases you have to look closely at the sensor chip itself. And the bottom image indeed looks like AHT20 and AHT21, respectively. The silk screen print does not really matter for what's on the board.

However, as long as you make sure to use most recent Tasmota, the driver is supposed to be able to handle the variations, when you use a build including `USE_AHT2x`.

When it comes to what works - my experience has resulted in my preference being to just not use any of those AHT sensors, has seen too many errors without a a clear pattern.

**Comment by xostia** [comment]
Dear sfromis for your answer... I understand what you say about AHT2x, but something is different in the boards, because the sensor itself is the same.

I use, in my Tasmota version, the AHT2x driver and it works perfect with all the "AHT2x" written boards but not with any of the "AHT Sensor" (I tested many from different sets with AHT20 and AHT21).

The difference that I detected is that chip, K72 (working board, AHT2x label) and K270 (working board, AHT Sensor label).

If anybody can give light to this... I add a good picture for help.

![AHT2xdude](https://github.com/arendst/Tasmota/assets/90287008/c64f6b80-f484-419f-b320-4d13e7b90cde)

**Reply by barbudor** [reply]
I have some suspicion on the chip marked K270
Where it is located I would assume a pair of N-channel MOSFET used for signal level adaptation

K72 should be a 2N7002DW : a pair of 2 x N-channel MOSFET => the right part for such level adaptation

I can't find anything about K270. 
If they used a wrong part here, then the board would be unusable.

I would suggest that you test those boards with a basic Arduino sketch to be sure that the problem is not on the hardware

At last, as you would use that board with an ESP that is 3V3, the level adapter is not reqiured. It should be possible to remove that part and strap the pins as below:

![image](https://github.com/arendst/Tasmota/assets/2996042/b408a680-1c81-468f-bcd0-b71ea218c6a5)

If you decide to try that, I would like to see a photo of PCB without the K270 chip before you solder wire to confirm

**Reply by sfromis** [reply]
The legend on that "K270" looks a bit odd. "K27" in one font size, and then "0" above "3" in a smaller font. Not clear if that "03" is part of the "K27", becoming "K2703", or something else is going on.....

On a board variant with AHT30, I'm seeing a very similar component in the same spot, labeled "702". and what looks like resistors in the center is "472" instead of "822". Some AHT10 which I unfortunately got a long time ago has "K27" and "103".

**Reply by barbudor** [reply]
Good point
K27 is an marking valid for 2N7002 from différé t manugacturers
03 could be a manufacturing tag like 2003?

As you can see from the diagram above, the part is fully symetrical and could be mounted 180° which is the case of the K72 on the photo

So not much clues here
Still testing with a basic Arduino or Micropython/Circuitpython sketch would give some hints

**Comment by xostia** [comment]
The unmodified sensor is there in Tasmota but no data is sent:

> 19:44:06.376 CMD: i2cscan
> 19:44:06.394 RSL: RESULT = {"I2CScan":"Device(s) found on bus1 at 0x38"}

I tested the "AHT sensor" board with the bridges as you propose but the behaviour is the same. It appears with i2cscan but no data is sent.

Here is the picture of the board without the K270.

![AHTwithoutK270](https://github.com/arendst/Tasmota/assets/90287008/a0d93f9d-e88f-40c4-97f1-85bb1d752875)

I found this information: [https://iot-kmutnb.github.io/blogs/sensors/aht2x/](url)

I will try to check the unmodified sensor with a simple program in python...

**Reply by xostia** [reply]
FOUND SOMETHING!

Replacing 0xB1 with 0xBE in xsns_63_AHT1x.ino the board labeled "AHT Sensor" untouched... it works!

> #define AHTX_CMD     0xBE   // 0xB1 for "AHT2x" labeled boards and 0xBE for "AHT Sensor" labeled boards

This was discussed here [#20163](https://github.com/arendst/Tasmota/discussions/20163) but there was no "agreement", probably because they were using different boards with AHT20/21.

It seems that there are, at least, two versions of these AHT boards... one answers to 0xB1 (labeled "AHTX") and other with 0xBE (labeled "AHT Sensor"). I do not understand how the CMD is modified in the electronics of the board (if somebody can explain... out of curiosity).

Any idea about an easy way to modify the code in xsns_63_AHT1x.ino to perform a check for both CMD to make it compatible with both board versions?

**Reply by xostia** [reply]
PROPOSAL

I was checking the Datasheets AHT20 and AHT21. In both, 0xBE is the initialization command (set the calibration).

I ignore why 0xB1 works with labeled "AHT2x" boards and not with "AHT Sensor" boards.

0xBE works perfect with BOTH, differente labeled, boards. Right now I only have these boards with AHT21 but, reading the datasheet, the modification should work with AHT20 and AHT21.

@barbudor / @sfromis  can you consider an update in xsns_63_AHT1x.ino for Tasmota?

#define AHTX_CMD 0xBE // 0xB1 only works for "AHT2x" labeled boards, 0xBE for all

Thank you

**Reply by barbudor** [reply]
I will double check the datasheets

**Reply by sfromis** [reply]
Part of the scenario is to make sure that it also works for AHT25.

**Comment by xostia** [comment]
Thank you all. At least, I could finish 3 photometers waiting for its AHT.

![FreeDSM](https://github.com/arendst/Tasmota/assets/90287008/34c29dd3-9ffa-41b4-b65b-ca0fdef9d4fd)

**Comment by radzik2024** [comment]
Hello,

I have a similar problem with AHT25 sensor, it is connected with Luatos esp32c3 - pin to pin (without any resistors).

I blocked driver USE_VEML6070 (by I2CDriver12 0) and activated the right one (I2CDriver43 1), but unfortunately Tasmota does not return any values.

I am using generic Tasmota 14.4.1 for ESP32 - I did not compile a special version inc I2C. 

What am I doing wrong?
Does the 14.4.1 version have the drivers for I2c pre-compiled by default, or should I do a dedicated compilation?
Should I block all drivers from I2CDriver1 to I2CDriver83 and activate only I2CDriver43 1? Very much clicking :-(

Please help - how to solve this problem?

**Reply by sfromis** [reply]
If you did not do a custom build to include the sensor driver, you are getting nowhere.
https://tasmota.github.io/docs/BUILDS/
This was already mentioned in my very first response in this thread.

When enabled, you should have the address in the output of command `I2CScan`, number 43 in the output of the command `I2CDriver` (it is not useful to "just" issue that commands before the driver number is included), and at boot time a message like `I2C: AHT2X found at 0x38`

**Reply by radzik2024** [reply]
Thank you for your reply Mr. @sfromis 

Uuu so it looks like none of the builds have the USE_AHT2x driver, so I'm left with my own compilation as a first step :-(

It's going to be hard, because I'm not that much of an expert, well hard I'm left to dive into the documentation ....

**Reply by sfromis** [reply]
No, it is not going to be "hard" or take "expertise", just a straightforward tool in your browser.
https://tasmota.github.io/docs/Compile-your-build/
TasmoCompiler is preobably the easiest.

**Reply by radzik2024** [reply]
Right. Maybe it's not very complicated, but you have to take into account maintenance, subsequent updates to Tasmota core itself, etc. so some time spent.

I've decided to change the sensor to a DS18B20, because what I can see, it's supported by default ([https://tasmota.github.io/docs/DS18x20/](https://tasmota.github.io/docs/DS18x20/)). I don't care about humidity measurement, I had a component on hand, so I thought I would use it “off the shelf”.

**Reply by nikito7** [reply]
Esp32  needs changes here:

```
// edit > include/tasmota_configurations_ESP32
//

#undef USE_SHT
#undef USE_VEML6070 // UV sensor with conflicting I2C address
#undef USE_AHT1x
#undef USE_AHT2x
#define USE_AHT2x
```

With Arduino Core 3.1.x pullup are mandatory. 
Internal pullup will not work...

**Comment by Jason2866** [comment]
Locking since the OP question / problem is solved.
